src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java

```diff
@@ -584,4 +588,32 @@ private NotificationMessage establishVolatility(FormSession session) {
         }
         return null;
     }
+
+    public BaseResponseBean advanceSessionWithEndpoint(MenuSession menuSession, String endpointId, @Nullable HashMap<String, String> endpointArgs)
+            throws Exception {
+        Endpoint endpoint = menuSession.getEndpoint(endpointId);
+        if (endpoint == null) {
+            throw new RuntimeException("No endpoint provided to navigateToEndpoint");
```

Comment:
NOTE: This will be the presumed path when an app is changed such that an old endpoint no longer exists, right? We probably want to provide a more meaningful semantic error in that case to let people know why the link they followed can't work.

Reply:
```suggestion
            throw new RuntimeException("This link does not exist. Your app may have changed so that the given link is no longer valid");
```
Would this be better ?

Reply:
New wording sounds much better, I updated the HQ PR to match it.